js/components/contextMenu.js

```diff
@@ -252,13 +237,165 @@ class ContextMenuSingle extends ImmutableComponent {
+          const newIndices = selectedIndex.slice()
+          newIndices.push(0)
+          windowActions.setContextMenuSelectedIndex(newIndices)
+
+          let openedSubmenuDetails = this.props.contextMenuDetail.get('openedSubmenuDetails')
+            ? this.props.contextMenuDetail.get('openedSubmenuDetails')
+            : new Immutable.List()
+
+          const rect = this.getContextMenuItemBounds()
+          const itemHeight = (rect.bottom - rect.top)
+
+          openedSubmenuDetails = openedSubmenuDetails.push(Immutable.fromJS({
+            y: (rect.top - itemHeight),
+            template: selectedMenuItem.get('submenu')
+          }))
+
+          windowActions.setContextMenuDetail(this.props.contextMenuDetail.set('openedSubmenuDetails', openedSubmenuDetails))
+        }
+        break
+
+      case keyCodes.UP:
+      case keyCodes.DOWN:
+        if (this.props.contextMenuDetail) {
+          const nextIndex = wrappingClamp(
+            currentIndex + (e.which === keyCodes.UP ? -1 : 1),
+            0,
+            this.maxIndex(selectedTemplate))
+
+          const newIndices = selectedIndex.slice()
+          newIndices[selectedIndex.length - 1] = nextIndex
+          windowActions.setContextMenuSelectedIndex(newIndices)
+        }
+        break
+    }
+  }
+
+  onKeyUp (e) {
+    e.preventDefault()
+  }
+
   onClick () {
     windowActions.resetMenuState()
   }
+
+  getTemplateItemsOnly (template) {
+    return template.filter((element) => {
+      if (element.get('type') === separatorMenuItem.type) return false
+      if (element.has('visible')) return element.get('visible')
+      return true
+    })
+  }
+
+  getMenuByIndex (selectedIndex, parentItem, currentDepth) {
+    parentItem = this.getTemplateItemsOnly(parentItem)
+    if (!currentDepth) currentDepth = 0
+
+    const selectedIndices = selectedIndex.slice(1)
+    if (selectedIndices.length === 0) return parentItem
+
+    const submenuIndex = selectedIndex[0]
+    const childItem = parentItem.get(submenuIndex)
+
+    if (childItem && childItem.get('submenu')) {
+      return this.getMenuByIndex(selectedIndices, childItem.get('submenu'), currentDepth + 1)
+    }
+
+    return parentItem
+  }
+
+  getContextMenuItemBounds () {
+    const selected = document.querySelectorAll('.contextMenuItem.selectedByKeyboard')
+    if (selected.length > 0) {
+      return selected.item(selected.length - 1).getBoundingClientRect()
+    }
+    return null
+  }
+
+  /**
+   * Upper bound for the active / focused menu.
+   */
+  maxIndex (template) {
+    return template.filter((element) => {
+      if (element.get('type') === separatorMenuItem.type) return false
+      if (element.has('visible')) return element.get('visible')
+      return true
+    }).size - 1
+  }
+
   get openedSubmenuDetails () {
     return this.props.contextMenuDetail.get('openedSubmenuDetails') || new Immutable.List()
   }
+
+  get hasSubmenuSelection () {
+    return this.props.selectedIndex.length > 1
```

Comment:
I think this line should be `(this.props.selectedIndex === null) ? false : this.props.selectedIndex.length > 1`
There is a chance `this.props.selectedIndex` will be `null` regarding https://github.com/brave/browser-laptop/pull/6452/files#diff-195859572336a7940350e808382f7fe8R398
STR:
![9co43ds6ku](https://cloud.githubusercontent.com/assets/11330831/21756257/d7c86f14-d659-11e6-9489-622f61b7348b.gif)